File: src/main/java/com/paragon/client/managers/StorageManager.java
package com.paragon.client.managers;

import com.paragon.Paragon;
import com.paragon.client.managers.alt.Alt;
import com.paragon.client.managers.social.Player;
import com.paragon.client.managers.social.Relationship;
import com.paragon.client.systems.module.hud.HUDModule;
import com.paragon.client.systems.module.setting.Bind;
import com.paragon.client.systems.module.setting.Setting;
import org.apache.commons.io.Charsets;
import org.apache.commons.io.FileUtils;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.awt.*;
import java.io.*;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * @author Wolfsurge
 * @since 2/2/22
 */
@SuppressWarnings("all") // FUCK warnings
public class StorageManager {

    public File mainFolder = new File("paragon");
    public File configFolder = new File("paragon/configs/");
    public File socialFolder = new File("paragon/social");

    @SuppressWarnings("all")
    public static JSONObject getJSON(File file) throws IOException, JSONException {
        return new JSONObject(FileUtils.readFileToString(file, Charsets.UTF_8));
    }

    public void saveModules(String configName) {
        // Create configs folder if it doesn't already exist
        if (!configFolder.exists()) {
            configFolder.mkdirs();
        }

        File saveConfigFolder = new File("paragon/configs/" + configName);

        // Create the folder if it doesn't exist
        if (!saveConfigFolder.exists()) {
            saveConfigFolder.mkdirs();
        }

        Paragon.INSTANCE.getModuleManager().getModules().forEach(module -> {
            try {
                JSONObject jsonObject = new JSONObject();

                jsonObject.put("enabled", module.isEnabled());

                if (module instanceof HUDModule) {
                    jsonObject.put("x", ((HUDModule) module).getX());
                    jsonObject.put("y", ((HUDModule) module).getY());
                }

                for (Setting<?> setting : module.getSettings()) {
                    if (setting.getValue() instanceof Color) {
                        jsonObject.put(setting.getName(), ((Color) setting.getValue()).getRed() + ":" + ((Color) setting.getValue()).getGreen() + ":" + ((Color) setting.getValue()).getBlue() + ":" + setting.getAlpha() + ":" + setting.isRainbow() + ":" + setting.getRainbowSpeed() + ":" + setting.getRainbowSaturation() + ":" + setting.isSync());
                    } else if (setting.getValue() instanceof Bind) {
                        jsonObject.put(setting.getName(), ((Bind) setting.getValue()).getButtonCode() + ":" + ((Bind) setting.getValue()).getDevice());
                    } else {
                        jsonObject.put(setting.getName(), setting.getValue());
                    }

                    if (!setting.getSubsettings().isEmpty()) {
                        for (Setting<?> subsetting : setting.getSubsettings()) {
                            String subsettingName = subsetting.getParentSetting().getName() + " " + subsetting.getName();

                            if (subsetting.getValue() instanceof Color) {
                                jsonObject.put(subsettingName, ((Color) subsetting.getValue()).getRed() + ":" + ((Color) subsetting.getValue()).getGreen() + ":" + ((Color) subsetting.getValue()).getBlue() + ":" +  subsetting.getAlpha() + ":" + subsetting.isRainbow() + ":" + subsetting.getRainbowSpeed() + ":" + subsetting.getRainbowSaturation() + ":" + subsetting.isSync());
                            } else if (subsetting.getValue() instanceof Bind) {
                                jsonObject.put(subsettingName, ((Bind) subsetting.getValue()).getButtonCode() + ":" + ((Bind) subsetting.getValue()).getDevice());
                            } else {
                                jsonObject.put(subsettingName, subsetting.getValue());
                            }
                        }
                    }
                }

                // Write to file
                FileWriter fileWriter = new FileWriter(new File(saveConfigFolder, module.getName() + ".json"));
                fileWriter.write(jsonObject.toString(4));
                fileWriter.flush();
                fileWriter.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        });
    }

    public void loadModules(String configName) {
        File loadFolder = new File("paragon/configs/" + configName);

        if (!loadFolder.exists()) {
            loadFolder.mkdirs();
        }

        try {
            Paragon.INSTANCE.getModuleManager().getModules().forEach(module -> {
                try {
                    JSONObject moduleJSON = getJSON(new File(loadFolder, module.getName() + ".json"));

                    if (moduleJSON.has("x") && moduleJSON.has("y")) {
                        ((HUDModule) module).setX(moduleJSON.getInt("x"));
                        ((HUDModule) module).setY(moduleJSON.getInt("y"));
                    }

                    for (Setting<?> setting : module.getSettings()) {
                        try {
                            if (setting.getValue() instanceof Boolean) {
                                ((Setting<Boolean>) setting).setValue(moduleJSON.getBoolean(setting.getName()));
                            }

                            else if (setting.getValue() instanceof Bind) {
                                String[] parts = moduleJSON.getString(setting.getName()).split(":");

                                ((Bind) setting.getValue()).setButtonCode(Integer.parseInt(parts[0]));
                                ((Bind) setting.getValue()).setDevice(Enum.valueOf(Bind.Device.class, parts[1]));
                            }

                            else if (setting.getValue() instanceof Float) {
                                ((Setting<Float>) setting).setValue(moduleJSON.getFloat(setting.getName()));
                            }

                            else if (setting.getValue() instanceof Double) {
                                ((Setting<Double>) setting).setValue(moduleJSON.getDouble(setting.getName()));
                            }

                            else if (setting.getValue() instanceof Enum<?>) {
                                Enum<?> value = Enum.valueOf(((Enum) setting.getValue()).getClass(), moduleJSON.getString(setting.getName()));

                                int i = 0;
                                for (Enum<?> enumValue : ((Enum<?>) setting.getValue()).getClass().getEnumConstants()) {
                                    if (enumValue.name().equals(value.name())) {
                                        setting.setIndex(i);
                                        break;
                                    }

                                    i++;
                                }

                                ((Setting<Enum<?>>) setting).setValue(value);
                            }

                            else if (setting.getValue() instanceof Color) {
                                String[] values = moduleJSON.getString(setting.getName()).split(":");

                                Color colour = new Color(Integer.parseInt(values[0]) / 255f, Integer.parseInt(values[1]) / 255f, Integer.parseInt(values[2]) / 255f, Float.parseFloat(values[3]) / 255f);

                                ((Setting<Color>) setting).setValue(colour);
                                setting.setRainbow(Boolean.parseBoolean(values[4]));
                                setting.setRainbowSpeed(Float.parseFloat(values[5]));
                                setting.setRainbowSaturation(Float.parseFloat(values[6]));
                                setting.setSync(Boolean.parseBoolean(values[7]));
                            }
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                        
                        for (Setting<?> subsetting : setting.getSubsettings()) {
                            String subsettingName = setting.getName() + " " + subsetting.getName();

                            try {
                                if (subsetting.getValue() instanceof Boolean) {
                                    ((Setting<Boolean>) subsetting).setValue(moduleJSON.getBoolean(subsettingName));
                                }

                                else if (subsetting.getValue() instanceof Bind) {
                                    String[] parts = moduleJSON.getString(subsettingName).split(":");

                                    ((Bind) subsetting.getValue()).setButtonCode(Integer.parseInt(parts[0]));
                                    ((Bind) subsetting.getValue()).setDevice(Enum.valueOf(Bind.Device.class, parts[1]));
                                }

                                else if (subsetting.getValue() instanceof Float) {
                                    ((Setting<Float>) subsetting).setValue(moduleJSON.getFloat(subsettingName));
                                }

                                else if (subsetting.getValue() instanceof Double) {
                                    ((Setting<Double>) subsetting).setValue(moduleJSON.getDouble(subsettingName));
                                }

                                else if (subsetting.getValue() instanceof Enum<?>) {
                                    Enum<?> value = Enum.valueOf(((Enum) subsetting.getValue()).getClass(), moduleJSON.getString(subsettingName));

                                    int i = 0;
                                    for (Enum<?> enumValue : ((Enum<?>) subsetting.getValue()).getClass().getEnumConstants()) {
                                        if (enumValue.name().equals(value.name())) {
                                            subsetting.setIndex(i);
                                            break;
                                        }

                                        i++;
                                    }

                                    ((Setting<Enum<?>>) subsetting).setValue(value);
                                }

                                else if (subsetting.getValue() instanceof Color) {
                                    String[] values = moduleJSON.getString(subsettingName).split(":");

                                    Color colour = new Color(Integer.parseInt(values[0]) / 255f, Integer.parseInt(values[1]) / 255f, Integer.parseInt(values[2]) / 255f, Float.parseFloat(values[3]) / 255f);

                                    ((Setting<Color>) subsetting).setValue(colour);
                                    subsetting.setRainbow(Boolean.parseBoolean(values[4]));
                                    subsetting.setRainbowSpeed(Float.parseFloat(values[5]));
                                    subsetting.setRainbowSaturation(Float.parseFloat(values[6]));
                                    subsetting.setSync(Boolean.parseBoolean(values[7]));
                                }
                            } catch (Exception e) {
                                e.printStackTrace();
                            }
                        }
                    }

                    if (moduleJSON.getBoolean("enabled") == !module.isEnabled()) {
                        module.toggle();
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            });
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * Saves the social interactions
     */
    public void saveSocial() {
        // Create friends folder if it doesn't already exist
        if (!socialFolder.exists()) {
            socialFolder.mkdirs();
        }

        try {
            // Create new friends json
            File file = new File("paragon/social/social_interactions.json");

            // Create file
            file.createNewFile();

            // Create new JSON object
            JSONObject jsonObject = new JSONObject();

            // Create file writer
            FileWriter fileWriter = new FileWriter(file);

            try {
                // Create a new array (for player info)
                JSONArray array = new JSONArray();

                for (Player player : Paragon.INSTANCE.getSocialManager().players) {
                    // Put the player's info in the array - name:relationship
                    array.put(player.getName() + ":" + player.getRelationship().name());
                }

                // Add array to json object
                jsonObject.putOpt("acquaintances", array);

                // Write JSON to file
                fileWriter.write(jsonObject.toString(4));
            } catch (IOException e) {
                e.printStackTrace();
            }

            // Flush file writer
            fileWriter.flush();

            // Close file writer
            fileWriter.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * Loads the social interactions
     */
    public void loadSocial() {
        // Create friends folder if it doesn't already exist
        if (!socialFolder.exists()) {
            socialFolder.mkdirs();
        }

        try {
            // Load JSON
            JSONObject jsonObject = getJSON(new File("paragon/social/social_interactions.json"));

            // Load array
            JSONArray acquaintances = jsonObject.getJSONArray("acquaintances");

            // For every value in array, create a new player
            for (int i = 0; i < acquaintances.length(); i++) {
                String[] info = String.valueOf(acquaintances.get(i)).split(":");
                Player player = new Player(info[0], Relationship.valueOf(info[1]));
                Paragon.INSTANCE.getSocialManager().addPlayer(player);
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void saveAlts() {
        // Create main folder if it doesn't already exist
        if (!mainFolder.exists()) {
            mainFolder.mkdirs();
        }

        try {
            // Create new friends json
            File file = new File("paragon/alts.json");

            // Create file
            file.createNewFile();

            // Create new JSON object
            JSONObject jsonObject = new JSONObject();

            // Create file writer
            FileWriter fileWriter = new FileWriter(file);

            try {
                // Create a new array (for alt info)
                JSONArray array = new JSONArray();

                for (Alt alt : Paragon.INSTANCE.getAltManager().getAlts()) {
                    // Put the player's info in the array - email:password
                    array.put(alt.getEmail() + ":" + alt.getPassword());
                }

                // Add array to json object
                jsonObject.putOpt("alts", array);

                // Write JSON to file
                fileWriter.write(jsonObject.toString(4));
            } catch (IOException e) {
                e.printStackTrace();
            }

            // Flush file writer
            fileWriter.flush();

            // Close file writer
            fileWriter.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void loadAlts() {
        // Create main folder if it doesn't already exist
        if (!mainFolder.exists()) {
            mainFolder.mkdirs();
        }

        try {
            // Load JSON
            JSONObject jsonObject = getJSON(new File("paragon/alts.json"));

            // Load array
            JSONArray alts = jsonObject.getJSONArray("alts");

            // For every value in array, create a new player
            for (int i = 0; i < alts.length(); i++) {
                String[] info = String.valueOf(alts.get(i)).split(":");
                Alt alt = new Alt(info[0], info[1]);
                Paragon.INSTANCE.getAltManager().addAlt(alt);
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void saveOther() {
        if (!mainFolder.exists()) {
            mainFolder.mkdirs();
        }

        try {
            File file = new File("paragon/client.json");
            file.createNewFile();

            JSONObject jsonObject = new JSONObject();

            FileWriter fileWriter = new FileWriter(file);

            try {
                jsonObject.put("mainmenu", Paragon.INSTANCE.isParagonMainMenu());
                fileWriter.write(jsonObject.toString(4));
            } catch (IOException exception) {
                exception.printStackTrace();
            }

            fileWriter.flush();
            fileWriter.close();
        } catch (Exception exception) {
            exception.printStackTrace();
        }
    }

    public void loadOther() {
        if (!mainFolder.exists()) {
            mainFolder.mkdirs();
        }

        try {
            // Load JSON
            JSONObject jsonObject = getJSON(new File("paragon/client.json"));

            Paragon.INSTANCE.setParagonMainMenu(jsonObject.getBoolean("mainmenu"));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

}

File: src/main/kotlin/com/paragon/client/managers/StorageManager.kt
package com.paragon.client.managers

import com.paragon.Paragon
import com.paragon.api.setting.Bind
import com.paragon.api.setting.Bind.Device
import com.paragon.api.setting.Setting
import com.paragon.client.managers.alt.Alt
import com.paragon.client.managers.social.Player
import com.paragon.client.managers.social.Relationship
import com.paragon.client.systems.module.hud.HUDModule
import org.apache.commons.io.FileUtils
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.awt.Color
import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files

/**
 * @author SooStrator1136
 */
class StorageManager {

    private val mainFolder = File("paragon")
    private val configFolder = File("paragon${File.separator}configs${File.separator}")
    private val socialFolder = File("paragon${File.separator}social")

    @Throws(IOException::class, JSONException::class)
    private fun getJSON(file: File) = if (!Files.exists(file.toPath())) null else JSONObject(FileUtils.readFileToString(file, StandardCharsets.UTF_8))

    fun saveModules(configName: String) {
        // Create configs folder if it doesn't already exist
        if (!configFolder.exists()) {
            configFolder.mkdirs()
        }

        val configFile = File("paragon${File.separator}configs${File.separator}$configName.json")

        val json = JSONObject()

        Paragon.INSTANCE.moduleManager.modules.forEach {
            try {
                val settings = JSONObject()

                settings.put("enabled", it.isEnabled)

                if (it is HUDModule) {
                    settings.put("x", it.x)
                    settings.put("y", it.y)
                }

                it.settings.forEach {setting ->
                    try {
                        when (setting.value) {
                            is Color -> {
                                val color = setting.value as Color
                                settings.put(
                                    setting.name,
                                    color.red.toString() + ":" + color.green + ":" + color.blue + ":" + setting.alpha + ":" + setting.isRainbow + ":" + setting.rainbowSpeed + ":" + setting.rainbowSaturation + ":" + setting.isSync
                                )
                            }

                            is Bind -> {
                                val bind = setting.value as Bind
                                settings.put(
                                    setting.name,
                                    bind.buttonCode.toString() + ":" + bind.device
                                )
                            }

                            else -> settings.put(setting.name, setting.value)
                        }

                        if (setting.subsettings.isNotEmpty()) {
                            for (subSetting in setting.subsettings) {
                                val subSettingName = subSetting.parentSetting?.name + " " + subSetting.name
                                when (subSetting.value) {
                                    is Color -> {
                                        val color = subSetting.value as Color
                                        settings.put(
                                            subSettingName,
                                            color.red.toString() + ":" + color.green + ":" + color.blue + ":" + subSetting.alpha + ":" + subSetting.isRainbow + ":" + subSetting.rainbowSpeed + ":" + subSetting.rainbowSaturation + ":" + subSetting.isSync
                                        )
                                    }

                                    is Bind -> {
                                        val bind = subSetting.value as Bind
                                        settings.put(
                                            subSettingName,
                                            bind.buttonCode.toString() + ":" + bind.device
                                        )
                                    }

                                    else -> settings.put(subSettingName, subSetting.value)
                                }
                            }
                        }
                    } catch (e: JSONException) {
                        e.printStackTrace()
                    }
                }

                json.put(it.name, settings)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }

        try {
            // Create file writer instance
            val fileWriter = FileWriter(configFile)

            // Write with indentation factor of 4
            fileWriter.write(json.toString(4))

            // Flush and close
            fileWriter.flush()
            fileWriter.close()
        } catch (e: JSONException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        /* val saveConfigFolder = File("paragon${File.separator}configs${File.separator}$configName")

        // Create the folder if it doesn't exist
        if (!saveConfigFolder.exists()) {
            saveConfigFolder.mkdirs()
        }

        Paragon.INSTANCE.moduleManager.modules.forEach { module ->
            try {
                val jsonObject = JSONObject()

                jsonObject.put("enabled", module.isEnabled)

                if (module is HUDModule) {
                    jsonObject.put("x", module.x)
                    jsonObject.put("y", module.y)
                }

                for (setting in module.settings) {
                    when (setting.value) {
                        is Color -> {
                            val color = setting.value as Color
                            jsonObject.put(
                                setting.name,
                                color.red.toString() + ":" + color.green + ":" + color.blue + ":" + setting.alpha + ":" + setting.isRainbow + ":" + setting.rainbowSpeed + ":" + setting.rainbowSaturation + ":" + setting.isSync
                            )
                        }

                        is Bind -> {
                            val bind = setting.value as Bind
                            jsonObject.put(
                                setting.name,
                                bind.buttonCode.toString() + ":" + bind.device
                            )
                        }

                        else -> jsonObject.put(setting.name, setting.value)
                    }

                    if (setting.subsettings.isNotEmpty()) {
                        for (subSetting in setting.subsettings) {
                            val subSettingName = subSetting.parentSetting?.name + " " + subSetting.name
                            when (subSetting.value) {
                                is Color -> {
                                    val color = subSetting.value as Color
                                    jsonObject.put(
                                        subSettingName,
                                        color.red.toString() + ":" + color.green + ":" + color.blue + ":" + subSetting.alpha + ":" + subSetting.isRainbow + ":" + subSetting.rainbowSpeed + ":" + subSetting.rainbowSaturation + ":" + subSetting.isSync
                                    )
                                }

                                is Bind -> {
                                    val bind = subSetting.value as Bind
                                    jsonObject.put(
                                        subSettingName,
                                        bind.buttonCode.toString() + ":" + bind.device
                                    )
                                }

                                else -> jsonObject.put(subSettingName, subSetting.value)
                            }
                        }
                    }
                }

                // Write to file
                FileWriter(File(saveConfigFolder, module.name + ".json")).use {
                    it.write(jsonObject.toString(4))
                    it.flush()
                }
            } catch (throwable: Throwable) {
                throwable.printStackTrace()
            }
        } */
    }

    @Suppress("UNCHECKED_CAST")
    fun loadModules(configName: String) {
        // Create configs folder if it doesn't already exist
        if (!configFolder.exists()) {
            configFolder.mkdirs()
        }

        val configFile = File("paragon${File.separator}configs${File.separator}$configName.json")

        val json = getJSON(configFile) ?: return

        Paragon.INSTANCE.moduleManager.modules.forEach {
            try {
                val settings = json.getJSONObject(it.name)

                if (settings.has("x") && settings.has("y")) {
                    it as HUDModule
                    it.x = settings.getInt("x").toFloat()
                    it.y = settings.getInt("y").toFloat()
                }

                fun loadSetting(setting: Setting<*>, isSub: Boolean) {
                    runCatching {
                        val settingName = if (isSub) setting.parentSetting?.name + " " + setting.name else setting.name

                        when (setting.value) {
                            is Boolean -> (setting as Setting<Boolean?>).setValue(settings.getBoolean(settingName))
                            is Bind -> {
                                val bind = setting.value as Bind
                                val parts = settings.getString(settingName).split(":".toRegex()).toTypedArray()

                                bind.buttonCode = parts[0].toInt()
                                bind.device = java.lang.Enum.valueOf(
                                    Device::class.java,
                                    parts[1]
                                )
                            }

                            is Float -> (setting as Setting<Float?>).setValue(settings.getFloat(settingName))
                            is Double -> (setting as Setting<Double?>).setValue(settings.getDouble(settingName))
                            is Enum<*> -> {
                                val enum = setting.value as Enum<*>
                                val value = java.lang.Enum.valueOf(enum::class.java, settings.getString(settingName))

                                run breakLoop@{
                                    enum::class.java.enumConstants.forEachIndexed { index, enumValue ->
                                        if (enumValue.name == value.name) {
                                            setting.index = index
                                            return@breakLoop
                                        }
                                    }
                                }

                                (setting as Setting<Enum<*>>).setValueRaw(value)
                            }

                            is Color -> {
                                val values = settings.getString(settingName).split(":".toRegex()).toTypedArray()

                                val color = Color(
                                    values[0].toInt() / 255f,
                                    values[1].toInt() / 255f,
                                    values[2].toInt() / 255f,
                                    values[3].toFloat() / 255f
                                )

                                setting.alpha = values[3].toFloat()
                                setting.isRainbow = java.lang.Boolean.parseBoolean(values[4])
                                setting.rainbowSpeed = values[5].toFloat()
                                setting.rainbowSaturation = values[6].toFloat()
                                setting.isSync = java.lang.Boolean.parseBoolean(values[7])
                                (setting as Setting<Color?>).setValue(color)
                            }
                        }
                    }
                }

                it.settings.forEach { setting ->
                    loadSetting(setting, false)

                    setting.subsettings.forEach { subSetting ->
                        loadSetting(subSetting, true)
                    }
                }

                if (settings.getBoolean("enabled") == !it.isEnabled) {
                    it.toggle()
                }
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }

        /* val loadFolder = File("paragon${File.separator}configs${File.separator}$configName")

        if (!loadFolder.exists()) {
            loadFolder.mkdirs()
            return
        }

        Paragon.INSTANCE.moduleManager.modules.forEach { module ->
            val moduleJSON = getJSON(File(loadFolder, module.name + ".json")) ?: return@forEach

            try {
                if (moduleJSON.has("x") && moduleJSON.has("y")) {
                    module as HUDModule
                    module.x = moduleJSON.getInt("x").toFloat()
                    module.y = moduleJSON.getInt("y").toFloat()
                }

                fun loadSetting(setting: Setting<*>, isSub: Boolean) {
                    runCatching {
                        val settingName = if (isSub) setting.parentSetting?.name + " " + setting.name else setting.name

                        when (setting.value) {
                            is Boolean -> (setting as Setting<Boolean?>).setValue(moduleJSON.getBoolean(settingName))
                            is Bind -> {
                                val bind = setting.value as Bind
                                val parts = moduleJSON.getString(settingName).split(":".toRegex()).toTypedArray()

                                bind.buttonCode = parts[0].toInt()
                                bind.device = java.lang.Enum.valueOf(
                                    Device::class.java,
                                    parts[1]
                                )
                            }

                            is Float -> (setting as Setting<Float?>).setValue(moduleJSON.getFloat(settingName))
                            is Double -> (setting as Setting<Double?>).setValue(moduleJSON.getDouble(settingName))
                            is Enum<*> -> {
                                val enum = setting.value as Enum<*>
                                val value = java.lang.Enum.valueOf(
                                    enum::class.java,
                                    moduleJSON.getString(settingName)
                                )

                                run breakLoop@{
                                    enum::class.java.enumConstants.forEachIndexed { index, enumValue ->
                                        if (enumValue.name == value.name) {
                                            setting.index = index
                                            return@breakLoop
                                        }
                                    }
                                }

                                (setting as Setting<Enum<*>>).setValue(value)
                            }

                            is Color -> {
                                val values = moduleJSON.getString(settingName).split(":".toRegex()).toTypedArray()

                                val color = Color(
                                    values[0].toInt() / 255f,
                                    values[1].toInt() / 255f,
                                    values[2].toInt() / 255f,
                                    values[3].toFloat() / 255f
                                )

                                setting.alpha = values[3].toFloat()
                                setting.isRainbow = java.lang.Boolean.parseBoolean(values[4])
                                setting.rainbowSpeed = values[5].toFloat()
                                setting.rainbowSaturation = values[6].toFloat()
                                setting.isSync = java.lang.Boolean.parseBoolean(values[7])
                                (setting as Setting<Color?>).setValue(color)
                            }
                        }
                    }
                }

                module.settings.forEach {
                    loadSetting(it, false)
                    it.subsettings.forEach { subSetting ->
                        loadSetting(subSetting, true)
                    }
                }

                if (moduleJSON.getBoolean("enabled") == !module.isEnabled) {
                    module.toggle()
                }
            } catch (throwable: Throwable) {
                throwable.printStackTrace()
            }
        } */
    }

    /**
     * Saves the social interactions
     */
    fun saveSocial() {
        if (!socialFolder.exists()) {
            socialFolder.mkdir() //Create friends folder if it doesn't already exist
        }

        val file = File("paragon${File.separator}social${File.separator}social_interactions.json")
        file.createNewFile()

        FileWriter(file).use { writer ->
            val jsonObject = JSONObject()
            val array = JSONArray()

            Paragon.INSTANCE.socialManager.players.forEach {
                array.put(it.name + ":" + it.relationship.name) //Put the player's info in the array - name:relationship
            }

            jsonObject.putOpt("acquaintances", array) // Add array to json object

            writer.write(jsonObject.toString(4))
            writer.flush()
        }
    }

    /**
     * Loads the social interactions
     */
    fun loadSocial() {
        if (!socialFolder.exists()) {
            socialFolder.mkdirs() //Create friends folder if it doesn't already exist
            return //If there is no folder there won't be anything to load either
        }

        runCatching {
            val json =
                getJSON(File("paragon${File.separator}social${File.separator}social_interactions.json")) ?: return

            val array = json.getJSONArray("acquaintances")

            //For every value in array, create add a player to the SocialManager
            for (i in 0 until array.length()) {
                val info: Array<String> = array.get(i).toString().split(":".toRegex()).toTypedArray()
                Paragon.INSTANCE.socialManager.addPlayer(Player(info[0], Relationship.valueOf(info[1])))
            }
        }
    }

    fun saveAlts() {
        if (!mainFolder.exists()) {
            mainFolder.mkdirs() //Create main folder if it doesn't already exist
        }

        val file = File("paragon${File.separator}alts.json")
        file.createNewFile()

        val jsonObject = JSONObject()

        FileWriter(file).use { writer ->
            val array = JSONArray()

            Paragon.INSTANCE.altManager.alts.forEach {
                array.put(it.email + ":" + it.password) //Put the player's info in the array - email:password
            }

            jsonObject.putOpt("alts", array)

            writer.write(jsonObject.toString(4))
            writer.flush()
        }
    }

    fun loadAlts() {
        if (!mainFolder.exists()) {
            mainFolder.mkdirs()
            return
        }

        runCatching {
            val json = getJSON(File("paragon${File.separator}alts.json")) ?: return
            val alts = json.getJSONArray("alts")

            //For every entry in the array, add a new alt
            for (i in 0 until alts.length()) {
                val info = alts[i].toString().split(":".toRegex()).toTypedArray()
                Paragon.INSTANCE.altManager.addAlt(Alt(info[0], info[1]))
            }
        }
    }

    fun saveOther() {
        if (!mainFolder.exists()) {
            mainFolder.mkdirs()
        }

        runCatching {
            val file = File("paragon${File.separator}client.json")
            file.createNewFile()

            FileWriter(file).use { writer ->
                val jsonObject = JSONObject()

                jsonObject.put("mainmenu", Paragon.INSTANCE.isParagonMainMenu)

                var prefixes = ""
                for (prefix in Paragon.INSTANCE.commandManager.commonPrefixes) {
                    prefixes += "$prefix "
                }
                jsonObject.put("ignored_prefixes", prefixes)

                writer.write(jsonObject.toString(4))
                writer.flush()
            }
        }
    }

    fun loadOther() {
        if (!mainFolder.exists()) {
            mainFolder.mkdirs()
            return
        }

        runCatching {
            val jsonObject = getJSON(File("paragon${File.separator}client.json")) ?: return

            Paragon.INSTANCE.isParagonMainMenu = jsonObject.getBoolean("mainmenu")

            if (jsonObject.has("ignored_prefixes")) {
                for (prefix in jsonObject.getString("ignored_prefixes").toString().split(" ".toRegex())) {
                    Paragon.INSTANCE.commandManager.commonPrefixes.add(prefix)
                }
            }
        }
    }

}